RandBLAS/trig_skops.hh

```diff
@@ -0,0 +1,493 @@
+            for (int i = 0; i < n; ++i) {
+                // Set the counter for each random number (unique per thread)
+                c[0] = ctr_seed + i; // Ensure the counter is unique across threads
+
+                // Generate a 2x32-bit random number using the Philox generator
+                RNG::ctr_type r = rng(c, k);
+
+                // Convert the random number into a float in [0, 1) using u01fixedpt
+                float randValue = r123::u01fixedpt<float>(r.v[0]);
+
+                // Convert the float into a Rademacher entry (-1 or 1)
+                buff[i] = randValue < 0.5 ? -1 : 1;
+            }
+        }
+    }
+
+    std::vector<int64_t> generateRademacherVector_parallel(int64_t n) {
+        std::vector<int64_t> rademacherVec(n);
+
+        #pragma omp parallel
+        {
+            // Each thread gets its own random number generator
+            std::random_device rd;
+            std::mt19937 gen(rd());
+            std::bernoulli_distribution bernoulli(0.5);
+
+            #pragma omp for
+            for (int64_t i = 0; i < n; ++i) {
+                rademacherVec[i] = bernoulli(gen) ? 1 : -1;
+            }
+        }
+
+        return rademacherVec;
+    }
+
+    template<typename T>
+    void applyDiagonalRademacher(int64_t rows, int64_t cols, T* A) {
+        std::vector<int64_t> diag = generateRademacherVector_parallel(cols);
+
+        for(int col=0; col < cols; col++) {
+        if(diag[col] > 0)
+            continue;
+        blas::scal(rows, diag[col], &A[col * rows], 1);
+    }
+    }
+
+    template<typename T, SignedInteger sint_t>
+    void applyDiagonalRademacher(
+                                bool left,
+                                blas::Layout layout,
+                                int64_t rows,
+                                int64_t cols,
+                                T* A,
+                                sint_t* diag
+                                )
+    {
+        if(left && layout == blas::Layout::ColMajor) {
+            for(int col=0; col < cols; col++) {
+                if(diag[col] > 0)
+                    continue;
+                blas::scal(rows, diag[col], &A[col * rows], 1);
+            }
+        }
+        else if(left && layout == blas::Layout::RowMajor) {
+            for(int col=0; col < cols; col++) {
+                if(diag[col] > 0)
+                    continue;
+                blas::scal(rows, diag[col], &A[col], cols);
+            }
+        }
+        else if(!left && layout == blas::Layout::ColMajor) {
+            for(int row = 0; row < rows; row++) {
+                if(diag[row] > 0)
+                    continue;
+                blas::scal(cols, diag[row], &A[row], rows);
+            }
+        }
+        else {
+            for(int row = 0; row < rows; row++) {
+                if(diag[row] > 0)
+                    continue;
+                blas::scal(cols, diag[row], &A[row * cols], 1);
+            }
+        }
+    }
```

Comment:
This is fine for now, but it can and should be much more efficient. It's also probably better suited to RandBLAS/util.hh.